### ISSUE TITLE
[experimental] Catching errors when dispatching

### ISSUE THREAD

**Issue by frankdilo** [BendingSpoons/katana-swift#160]
### Problem

In the new experimental branch `dispatch` returns a promise that always resolves. I understand this is done to avoid having to do this `dispatch(SomeSideEffect()).then({})` when dispatching in most cases.

However, always resolving can weaken the error-handling capabilities offered by promises, especially when a `SideEffect` calls a throwing method of a manager.

### Example

Let me present an example. I have a MediaManager in my app that loads assets from the camera roll and some related side effects.

```swift
final class MediaManager {
  func loadAssets(_ identifiers: [String]) -> Promise<PHAsset> {
  ...
  }
}

struct LoadAssets: SideEffect {
  var identifiers: [String]
  func sideEffect(_ context: SideEffectContext<AppState, DependenciesContainer>) throws {
    // call the manager, among other things
    async { _ in
      let assets = try dependencies.mediaManager.loadAssets(self.identifiers) // 🛠
    }
  }
}

// this some other action called when the user taps a button
struct LoadSelectedAssets: SideEffect {
  func sideEffect(_ context: SideEffectContext<AppState, DependenciesContainer>) throws {
    context.dispatch(LoadAssets(identifiers: context.getState().selectedAssetsIdentifiers)) // 🚨
  }
}
```

It would be more natural to handle the error on the outer side effect (🚨), but we can't since the error is lost when `dispatch` auto-resolves. 

Another issue I see on line 🛠 is that since `sideEffect` throws, but the error is not propagated to the `dispatch` call site, the error that could arise from `ProjectManager.loadAssets` is never handled and we get no compile-time error.

### Possible solution

How about having a throwing variant of `dispatch` that propagates the error if the sideEffect throws an error?

**Comment by fonesti** [comment]
I think this is a legacy issue.
The error is lost only when using an "old" `Middleware` via `middlewareToInterceptor(_:)`.
I'm trying to fix the in #166

**Comment by bolismauro** [comment]
Fixed
